### PR TITLE
Introduce a store, store wn

### PR DESCRIPTION
**Pull request by mfine**
Introduce a store in the converter. Store the week number.

The store will be able to handle future things that require local storage like lock timer conversions.

/cc @ljbade

### CODE REVIEW

**Comment by ljbade** [comment]
Oh good idea for being able to store the lock counter

**Comment by mfine** [comment]
This is now susceptible to week number rollover.

**Comment by ljbade** [comment]
What do you mean by that?

**Comment by mfine** [comment]
@ljbade the previous code was recalculating the week number off the local clock for every observation message. This change now initializes the converter with a week number. But the week number won't be updated anymore on each observation. Instead, it will need to get updated when the tow rolls over. Does that sound right? I'm still unsure of where to grab the initial week number - the calculation off the local clock still seems less than robust. And unsure of how to calculate the tow rollover.

**Comment by ljbade** [comment]
@mfine Ah I see, I will have a good think about this and can pick it up next week since I will get a chance to enable 1013 on Geo++.

**Comment by ejconlon** [comment]
LGTM

**Comment by ljbade** [comment]
@mookerji You have the laptop with the Geo++ login? to see if we can enable 1013?

**Comment by mfine** [comment]
Rebased and integrated with the new L2 observations. @mookerji mind reviewing and checking out if everything looks ok?

File: haskell/src/Data/RTCM3/SBP.hs [hunk 1]
```diff
@@ -372,22 +379,29 @@ fromMsg1006 m =
...
 -- | Convert an RTCM message into possibly multiple SBP messages.
 --
-convert :: MonadIO m => RTCM3Msg -> m (Maybe [SBPMsg])
+convert :: MonadStore e m => RTCM3Msg -> m [SBPMsg]
 convert = \case
   (RTCM3Msg1002 m _rtcm3) -> do
     let sender = m ^. msg1002_header ^. gpsObservationHeader_station
     m' <- fromMsg1002 m
-    return $ Just [SBPMsgObs m' $ toSBP m' $ toSender sender]
+    return $ flip fmap m' $ \x -> SBPMsgObs x $ toSBP x $ toSender sender
```
Comment (ejconlon) [inline haskell/src/Data/RTCM3/SBP.hs]: Use <$> for fmap and remove the lambda?

File: haskell/src/Data/RTCM3/SBP/Types.hs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    , Applicative
+    , Monad
+    , MonadIO
+    , MonadTrans
+    )
+
+instance MonadThrow m => MonadThrow (ConvertT r m) where
+    throwM = lift . throwM
+
+instance MonadResource m => MonadResource (ConvertT r m) where
```
Comment (ejconlon) [inline haskell/src/Data/RTCM3/SBP/Types.hs]: idk if you need this if you remove the resourcet layer

File: haskell/src/Data/RTCM3/SBP/Types.hs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  liftResourceT = lift . liftResourceT
+
+instance Monad m => MonadReader e (ConvertT e m) where
+    ask     = ConvertT ask
+    local f = ConvertT . local f . unConvertT
+    reader  = ConvertT . reader
+
+instance MonadTransControl (ConvertT r) where
```
Comment (ejconlon) [inline haskell/src/Data/RTCM3/SBP/Types.hs]: what are this and monadbase used for?

File: haskell/src/Data/RTCM3/SBP/Types.hs (hunk 2 above)
Comment (mfine) [inline haskell/src/Data/RTCM3/SBP/Types.hs]: Still being used in the tests where there's sourcing from a file, which does have a `MonadResource` constraint:

```
/Users/mark/repos/work/gnss-converters/haskell/test/Test/Data/RTCM3/SBP.hs:36:5:
    No instance for (MonadResource (ConvertT Store (ResourceT IO)))
      arising from a use of ‘sourceFile’
    In the first argument of ‘(=$=)’, namely ‘sourceFile filename’
    In the first argument of ‘($$)’, namely
      ‘sourceFile filename =$= conduitDecode =$= CL.concatMapM convert’
    In the second argument of ‘($)’, namely
      ‘sourceFile filename =$= conduitDecode =$= CL.concatMapM convert
       $$ CL.consume’
```

Need to leave it in.

File: haskell/src/Data/RTCM3/SBP/Types.hs (hunk 3 above)
Comment (mfine) [inline haskell/src/Data/RTCM3/SBP/Types.hs]: Oh man, this is all the transformer boilerplate stuff - I'll see if any of it can be thinned.

**Comment by ejconlon** [comment]
LGTM

**Comment by mookerji** [comment]
👍 !!!!

**Pull request merged**